icu4c/source/common/uniset_props.cpp

```diff
@@ -220,438 +223,796 @@ UBool UnicodeSet::resemblesPattern(const UnicodeString& pattern, int32_t pos) {
-        // the set should be parsed.  If the '[' is part of the
-        // opening delimiter for this pattern, parse special
-        // strings "[", "[^", "[-", and "[^-".  Check for stand-in
-        // characters representing a nested set in the symbol
-        // table.
+  private:
+    // A version of getPos that returns its position instead of taking it as at out parameter, so we
+    // can have const positions.
+    RuleCharacterIterator::Pos getPos() const {
+        RuleCharacterIterator::Pos result;
+        chars_.getPos(result);
+        return result;
+    }
 
-        else {
-            // Prepare to backup if necessary
-            chars.getPos(backup);
-            c = chars.next(opts, literal, ec);
-            if (U_FAILURE(ec)) return;
-
-            if (c == u'[' && !literal) {
-                if (mode == 1) {
-                    chars.setPos(backup); // backup
-                    setMode = 1;
-                } else {
-                    // Handle opening '[' delimiter
-                    mode = 1;
-                    patLocal.append(u'[');
-                    chars.getPos(backup); // prepare to backup
-                    c = chars.next(opts, literal, ec); 
-                    if (U_FAILURE(ec)) return;
-                    if (c == u'^' && !literal) {
-                        invert = true;
-                        patLocal.append(u'^');
-                        chars.getPos(backup); // prepare to backup
-                        c = chars.next(opts, literal, ec);
-                        if (U_FAILURE(ec)) return;
+    LexicalElement nextToken() {
+        UErrorCode errorCode = U_ZERO_ERROR;
+        chars_.skipIgnored(charsOptions_);
+        if (chars_.atEnd()) {
+            return LexicalElement(LexicalElement::END_OF_TEXT, {}, getPos(), errorCode,
+                                  /*standIn=*/nullptr, u"");
+        }
+        const int32_t start = parsePosition_.getIndex();
+        const RuleCharacterIterator::Pos before = getPos();
+        // First try to get the next character without parsing escapes.
+        UBool unusedEscaped;
+        const UChar32 first =
+            chars_.next(charsOptions_ & ~RuleCharacterIterator::PARSE_ESCAPES, unusedEscaped, errorCode);
+        // '[', named-element, and property-query cannot be disabled by stand-in.
+        if (first == u'[' || first == u'\\') {
+            const RuleCharacterIterator::Pos afterFirst = getPos();
+            // This could be a property-query or named-element.
+            const UChar32 second = chars_.next(charsOptions_ & ~(RuleCharacterIterator::PARSE_ESCAPES |
+                                                                 RuleCharacterIterator::SKIP_WHITESPACE),
+                                               unusedEscaped, errorCode);
+            if ((first == u'[' && second == u':') ||
+                (first == u'\\' && (second == u'p' || second == u'P' || second == u'N'))) {
+                if (second == u'N') {
```

Comment:
Could you please split this function into somewhat smaller chunks?
For example, pull out something like `nextPropertyQueryToken(first, second, ...)` from here to just before `// Not a property-query.` ?

Reply:
Oops, I forgot about this one.

I will do that now.